File: quad_logger/scripts/plotState.m
function varargout = plotState(stateTraj,tWindow,varargin)
% plotState Plot state data for a trajectory.
%   plotState(stateTraj, tWindow) generates plots for each component of
%   the state message -- COM position, velocity, orientation, angular
%   velocity, joint states, and foot states -- for each time t =
%   [tWindow(1), tWindow(2)]. If tWindow = [], all data points are
%   displayed.
%
%   plotState(stateTraj, tWindow, lineStyle) plots all data with the line
%   style specified by lineStyle.
% 
%   plotState(stateTraj, tWindow, lineStyle, titles) adds figure titles if
%   titles == true.
%
%   plotState(stateTraj, tWindow, lineStyle, titles, figArray) plots data
%   from stateTraj on existing figure handles specified by figArray.
%
%   figArray = plotState(___) returns the handles of each figure contained
%   within figArray

% Use existing figure handles if requested
if (length(varargin)>=3) && ~isempty(varargin{3})
    COMTrajFig = varargin{3}(1);
    linearStateFig = varargin{3}(2);
    angularStateFig = varargin{3}(3);
    jointPositionFig = varargin{3}(4);
    jointVelocityFig = varargin{3}(5);
    jointEffortFig = varargin{3}(6);
    footPositionFig = varargin{3}(7);
    footVelocityFig = varargin{3}(8);
else
    h =  findobj('type','figure');
    n = length(h);
    
    COMTrajFig = n+1;
    linearStateFig = n+2;
    angularStateFig = n+3;
    jointPositionFig = n+4;
    jointVelocityFig = n+5;
    jointEffortFig = n+6;
    footPositionFig = n+7;
    footVelocityFig = n+8;
end

% Turn off the plot titles if requested
if (length(varargin)>=2)
    titles = varargin{2};
else
    titles = true;
end

% Set the line style if specified
if (length(varargin)>=1)
    lineStyle = varargin{1};
else
    lineStyle = '-';
end

% Set the time window if specified
if isempty(tWindow)
    traj_idx = 1:length(stateTraj.time);
else
    traj_idx = find(stateTraj.time>= tWindow(1) & stateTraj.time<= tWindow(2));
end

% Trim the data based on the requested window
stateTraj.time = stateTraj.time(traj_idx);
stateTraj.position = stateTraj.position(traj_idx,:);
stateTraj.velocity = stateTraj.velocity(traj_idx,:);
stateTraj.orientationRPY = stateTraj.orientationRPY(traj_idx,:);
stateTraj.orientationQuat = stateTraj.orientationQuat(traj_idx,:);
stateTraj.angularVelocity = stateTraj.angularVelocity(traj_idx,:);
stateTraj.jointPosition = stateTraj.jointPosition(traj_idx,:);
stateTraj.jointVelocity = stateTraj.jointVelocity(traj_idx,:);
stateTraj.jointEffort = stateTraj.jointEffort(traj_idx,:);
for i = 1:length(stateTraj.footPosition)
    stateTraj.footPosition{i} = stateTraj.footPosition{i}(traj_idx,:);
    stateTraj.footVelocity{i} = stateTraj.footVelocity{i}(traj_idx,:);
end

footColorVector = {[166,25,46]/255, [0,45,114]/255, [0,132,61]/255, [242,169,0]/255};
jointColorVector = footColorVector;
bodyColorVector = footColorVector;

num_feet = 4;

%% Plot spatial trajectory
COMTrajFig = figure(COMTrajFig);
COMTrajFig.Name = "com_trajectory";

hold on; view(3);
plot3(stateTraj.position(:,1), stateTraj.position(:,2), stateTraj.position(:,3), ...
    'Color', cmuColor('red-web'), 'LineStyle', lineStyle);
for i = 1:num_feet
   plot3(stateTraj.footPosition{i}(:,1), stateTraj.footPosition{i}(:,2), ...
       stateTraj.footPosition{i}(:,3), 'Color', footColorVector{i}, ...
       'LineStyle', lineStyle);
end
xlabel('X (m)')
ylabel('Y (m)')
zlabel('Z (m)')
if titles
    title('Body and Foot Trajectories')
end
axis equal

%% Plot linear state data
linearStateFig = figure(linearStateFig);
linearStateFig.Name = "linear_states";

subplot(2,1,1); hold on
for i = 1:3
    plot(stateTraj.time, stateTraj.position(:,i), 'Color', bodyColorVector{i},...
        'LineStyle', lineStyle);
end
ylabel('Position (m)')
if titles
    title('Linear States')
end
legend('X','Y','Z','location','East')
axis tight

subplot(2,1,2); hold on
for i = 1:3
    plot(stateTraj.time, stateTraj.velocity(:,i), 'Color', bodyColorVector{i},...
        'LineStyle', lineStyle);
end
ylabel('Velocity (m/s)')
xlabel('Time (s)')
axis tight
set(linearStateFig, 'Position', [100 100 1200 600])

align_Ylabels(linearStateFig);

%% Plot angular state data
angularStateFig = figure(angularStateFig);
angularStateFig.Name = "angular_states";

subplot(2,1,1); hold on
for i = 1:3
    plot(stateTraj.time, stateTraj.orientationRPY(:,i), 'Color', bodyColorVector{i},...
        'LineStyle', lineStyle);
end
ylabel('Ang. Pos. (rad)')
if titles
    title('Angular States')
end
legend('Roll','Pitch','Yaw','location','East')
axis([min(stateTraj.time), max(stateTraj.time), -pi, pi])

subplot(2,1,2); hold on
for i = 1:3
    plot(stateTraj.time, stateTraj.angularVelocity(:,i), 'Color', bodyColorVector{i},...
        'LineStyle', lineStyle);
end
ylabel('Ang. Vel. (rad/s)')
xlabel('Time (s)')
axis tight
set(angularStateFig, 'Position', [100 100 1200 600])

align_Ylabels(angularStateFig);

%% Plot joint positions

% Specify indices for leg joints
abIndex = [1:3:12];
hipIndex = [2:3:12];
kneeIndex = [3:3:12];
abadLim = [-1, 1];
hipLim = [-pi/2, pi];
kneeLim = [0, pi];

% Plot joint positions
jointPositionFig = figure(jointPositionFig);
jointPositionFig.Name = "joint_positions";

subplot(3,1,1); hold on;
for i = 1:num_feet
   plot(stateTraj.time, stateTraj.jointPosition(:,abIndex(i)), ...
       'Color', jointColorVector{i}, 'LineWidth', 2, 'LineStyle', lineStyle);
end
ylabel('Ab/Ad (rad)')
if titles
    title('Joint Angles')
end
axis([min(stateTraj.time), max(stateTraj.time), abadLim(1), abadLim(2)])

subplot(3,1,2); hold on;
for i = 1:num_feet
   plot(stateTraj.time, stateTraj.jointPosition(:,hipIndex(i)), ...
       'Color', jointColorVector{i}, 'LineWidth', 2, 'LineStyle', lineStyle);
end
ylabel('Hip (rad)')
legend('FL','BL','FR','BR','location','East')
axis([min(stateTraj.time), max(stateTraj.time), hipLim(1), hipLim(2)])

subplot(3,1,3); hold on;
for i = 1:num_feet
   plot(stateTraj.time, stateTraj.jointPosition(:,kneeIndex(i)), ...
       'Color', jointColorVector{i}, 'LineWidth', 2, 'LineStyle', lineStyle);
end
ylabel('Knee (rad)')
xlabel('Time (s)')
axis([min(stateTraj.time), max(stateTraj.time), kneeLim(1), kneeLim(2)])
set(jointPositionFig, 'Position', [100 100 1200 900])

align_Ylabels(jointPositionFig);

%% Plot joint velocities
abadVelLim = [-37.7, 37.7];
hipVelLim = [-37.7, 37.7];
kneeVelLim = [-25, 25];

jointVelocityFig = figure(jointVelocityFig);
jointVelocityFig.Name = "joint_velocities";

subplot(3,1,1); hold on;
for i = 1:num_feet
   plot(stateTraj.time, stateTraj.jointVelocity(:,abIndex(i)), ...
       'Color', jointColorVector{i}, 'LineWidth', 2, 'LineStyle', lineStyle);
end
ylabel('Ab/Ad (rad/s)')
if titles
    title('Joint Velocities')
end
axis([min(stateTraj.time), max(stateTraj.time), abadVelLim(1), abadVelLim(2)])

subplot(3,1,2); hold on;
for i = 1:num_feet
   plot(stateTraj.time, stateTraj.jointVelocity(:,hipIndex(i)), ...
       'Color', jointColorVector{i}, 'LineWidth', 2, 'LineStyle', lineStyle);
end
ylabel('Hip (rad/s)')
legend('FL','BL','FR','BR','location','East')
axis([min(stateTraj.time), max(stateTraj.time), hipVelLim(1), hipVelLim(2)])

subplot(3,1,3); hold on;
for i = 1:num_feet
   plot(stateTraj.time, stateTraj.jointVelocity(:,kneeIndex(i)), ...
       'Color', jointColorVector{i}, 'LineWidth', 2, 'LineStyle', lineStyle);
end
ylabel('Knee (rad/s)')
xlabel('Time (s)')
axis([min(stateTraj.time), max(stateTraj.time), kneeVelLim(1), kneeVelLim(2)])
set(jointVelocityFig, 'Position', [100 100 1200 900])

align_Ylabels(jointVelocityFig);

%% Plot estimate joint effort
abadEffLim = [-21, 21];
hipEffLim = [-21, 21];
kneeEffLim = [-32, 32];

jointEffortFig = figure(jointEffortFig);
jointEffortFig.Name = "joint_effort";

subplot(3,1,1); hold on;
for i = 1:num_feet
   plot(stateTraj.time, stateTraj.jointEffort(:,abIndex(i)), ...
       'Color', jointColorVector{i}, 'LineWidth', 2, 'LineStyle', lineStyle);
end
ylabel('Ab/Ad (A)')
if titles
    title('Joint Effort')
end
axis([min(stateTraj.time), max(stateTraj.time), abadEffLim(1), abadEffLim(2)])

subplot(3,1,2); hold on;
for i = 1:num_feet
   plot(stateTraj.time, stateTraj.jointEffort(:,hipIndex(i)), ...
       'Color', jointColorVector{i}, 'LineWidth', 2, 'LineStyle', lineStyle);
end
ylabel('Hip (A)')
legend('FL','BL','FR','BR','location','East')
axis([min(stateTraj.time), max(stateTraj.time), hipEffLim(1), hipEffLim(2)])

subplot(3,1,3); hold on;
for i = 1:num_feet
   plot(stateTraj.time, stateTraj.jointEffort(:,kneeIndex(i)), ...
       'Color', jointColorVector{i}, 'LineWidth', 2, 'LineStyle', lineStyle);
end
ylabel('Knee (A)')
xlabel('Time (s)')
axis([min(stateTraj.time), max(stateTraj.time), kneeEffLim(1), kneeEffLim(2)])
set(jointEffortFig, 'Position', [100 100 1200 900])

align_Ylabels(jointEffortFig);

%% Plot foot positions
footPositionFig = figure(footPositionFig);
footPositionFig.Name = "foot_positions";

subplot(3,1,1)
hold on;
for i = 1:num_feet
   plot(stateTraj.time, stateTraj.footPosition{i}(:,1), ...
       'Color', footColorVector{i}, 'LineWidth', 2, 'LineStyle', lineStyle);
end
ylabel('X (m)')
if titles
    title('Foot Positions')
end
% axis([min(stateTraj.time), max(stateTraj.time), -1, 4])
axis tight

subplot(3,1,2)
hold on;
for i = 1:num_feet
   plot(stateTraj.time, stateTraj.footPosition{i}(:,2), ...
       'Color', footColorVector{i}, 'LineWidth', 2, 'LineStyle', lineStyle);
end
ylabel('Y (m)')
% axis([min(stateTraj.time), max(stateTraj.time), -2 2])
axis tight
legend('FL','BL','FR','BR','location','East')

subplot(3,1,3)
hold on;
for i = 1:num_feet
   plot(stateTraj.time, stateTraj.footPosition{i}(:,3), ...
       'Color', footColorVector{i}, 'LineWidth', 2, 'LineStyle', lineStyle);
end
ylabel('Z (m)')
xlabel('Time (s)')
% axis([min(stateTraj.time), max(stateTraj.time), 0, 0.3])
axis tight
set(footPositionFig, 'Position', [100 100 1200 900])

align_Ylabels(footPositionFig);

%% Plot foot velocities
footVelocityFig = figure(footVelocityFig);
footVelocityFig.Name = "foot_velocities";

vel_axis = max(abs(cell2mat(stateTraj.footVelocity)), [], 'all');
if ~isempty(footVelocityFig.CurrentAxes)
    vel_axis = max(vel_axis, max(abs(footVelocityFig.CurrentAxes.YLim)));
end
subplot(3,1,1)
hold on;
for i = 1:num_feet
   plot(stateTraj.time, stateTraj.footVelocity{i}(:,1), ...
       'Color', footColorVector{i}, 'LineWidth', 2, 'LineStyle', lineStyle);
end
ylabel('X (m/s)')
if titles
    title('Foot Velocities')
end
axis([min(stateTraj.time), max(stateTraj.time), -vel_axis, vel_axis])

subplot(3,1,2)
hold on;
for i = 1:num_feet
   plot(stateTraj.time, stateTraj.footVelocity{i}(:,2), ...
       'Color', footColorVector{i}, 'LineWidth', 2, 'LineStyle', lineStyle);
end
ylabel('Y (m/s)')
axis([min(stateTraj.time), max(stateTraj.time), -vel_axis, vel_axis])
legend('FL','BL','FR','BR','location','East')

subplot(3,1,3)
hold on;
for i = 1:num_feet
   plot(stateTraj.time, stateTraj.footVelocity{i}(:,3), ...
       'Color', footColorVector{i}, 'LineWidth', 2, 'LineStyle', lineStyle);
end
ylabel('Z (m/s)')
xlabel('Time (s)')
axis([min(stateTraj.time), max(stateTraj.time), -vel_axis, vel_axis])
set(footVelocityFig, 'Position', [100 100 1200 900])

align_Ylabels(footVelocityFig);

%% Export

if nargout >0
    varargout{1} = [COMTrajFig, linearStateFig, angularStateFig, jointPositionFig, jointVelocityFig, ...
        jointEffortFig, footPositionFig, footVelocityFig];
end

File: quad_logger/scripts/plot_state.py
"""
State trajectory visualization module for quadruped robotics.

This module provides comprehensive visualization functionality for quadruped
robot state trajectories, including body motion (linear and angular states),
joint states (positions, velocities, efforts), and foot trajectories in 3D
space and over time.
"""

from math import pi
import numpy as np
import matplotlib.pyplot as plt


def plot_state(state_traj, t_window, *args):
    """
    Plot comprehensive state trajectory data for a quadruped robot.

    Creates multiple matplotlib figures showing body trajectories, linear and
    angular states, joint positions/velocities/efforts, and foot positions/
    velocities. All data is filtered by the specified time window.

    Args:
        state_traj (dict): State trajectory data containing:
            - "time" (np.ndarray): Time array
            - "position" (np.ndarray): Body position [x, y, z]
            - "velocity" (np.ndarray): Body linear velocity
            - "orientationRPY" (np.ndarray): Body orientation (roll, pitch, yaw)
            - "orientationQuat" (np.ndarray): Body orientation quaternion
            - "angularVelocity" (np.ndarray): Body angular velocity
            - "jointPosition" (np.ndarray): Joint angles for all legs
            - "jointVelocity" (np.ndarray): Joint velocities
            - "jointEffort" (np.ndarray): Joint efforts/currents
            - "footPosition" (list): Foot positions for each leg
            - "footVelocity" (list): Foot velocities for each leg
        t_window (list or np.ndarray): Time window for plotting [start, end].
            If empty, plots entire trajectory.
        *args: Variable length argument list:
            - args[0] (str, optional): Line style for plots (default: '-')
            - args[1] (bool, optional): Whether to show titles (default: True)
            - args[2] (list, optional): Existing figure handles to reuse,
              ordered as [com_traj, linear_state, angular_state,
              joint_position, joint_velocity, joint_effort, foot_position,
              foot_velocity]

    Returns:
        np.ndarray: Array of figure numbers for all created plots:
            [com_traj_fig, linear_state_fig, angular_state_fig,
             joint_position_fig, joint_velocity_fig, joint_effort_fig,
             foot_position_fig, foot_velocity_fig]

    Note:
        Not all state_traj keys are required. The function will only plot
        data for keys that exist in the dictionary. Reference trajectories
        may not include joint or foot data.
    """
    # Use existing figure handles if requested
    if len(args) >= 3 and len(args[2]) > 0:
        com_traj_fig = args[2][0]
        linear_state_fig = args[2][1]
        angular_state_fig = args[2][2]
        joint_position_fig = args[2][3]
        joint_velocity_fig = args[2][4]
        joint_effort_fig = args[2][5]
        foot_position_fig = args[2][6]
        foot_velocity_fig = args[2][7]
    else:
        h = plt.get_fignums()
        n = len(h)
        com_traj_fig = n + 1
        linear_state_fig = n + 2
        angular_state_fig = n + 3
        joint_position_fig = n + 4
        joint_velocity_fig = n + 5
        joint_effort_fig = n + 6
        foot_position_fig = n + 7
        foot_velocity_fig = n + 8

    titles = args[1] if len(args) >= 2 else True
    line_style = args[0] if len(args) >= 1 else '-'

    # Calculate time indices
    if len(t_window) == 0:
        traj_idx = np.arange(len(state_traj["time"]))
    else:
        traj_idx = np.where(
            (state_traj["time"] >= t_window[0]) &
            (state_traj["time"] <= t_window[1])
        )[0]

    # Only use keys that exist in the dictionary, some keys are not
    # included in reference trajectory
    keys_to_trim = ["time", "position", "velocity", "orientationRPY",
                    "orientationQuat", "angularVelocity", "jointPosition",
                    "jointVelocity", "jointEffort"]

    for key in keys_to_trim:
        if key in state_traj and len(state_traj[key]) > 0:
            state_traj[key] = state_traj[key][traj_idx]

    # Foot data
    if "footPosition" in state_traj:
        for i in range(len(state_traj["footPosition"])):
            state_traj["footPosition"][i] = (
                state_traj["footPosition"][i][traj_idx])
            state_traj["footVelocity"][i] = (
                state_traj["footVelocity"][i][traj_idx])

    # Color definitions
    foot_color_vector = [np.array([166, 25, 46]) / 255,
                         np.array([0, 45, 114]) / 255,
                         np.array([0, 132, 61]) / 255,
                         np.array([242, 169, 0]) / 255]
    body_color_vector = foot_color_vector

    num_feet = 4
    joint_color_vector = foot_color_vector

    # Plot 1: 3D Trajectory Plot
    fig1 = plt.figure(com_traj_fig)
    ax = plt.subplot(projection='3d')
    fig1.canvas.manager.set_window_title("body_and_foot_trajectories")

    if "position" in state_traj:
        ax.plot(
            state_traj["position"][:, 0],
            state_traj["position"][:, 1],
            state_traj["position"][:, 2],
            linestyle=line_style, label='Body'
        )

    if "footPosition" in state_traj:
        for i in range(len(state_traj["footPosition"])):
            ax.plot(
                state_traj["footPosition"][i][:, 0],
                state_traj["footPosition"][i][:, 1],
                state_traj["footPosition"][i][:, 2],
                color=foot_color_vector[i % 4],
                linestyle=line_style
            )

    ax.set_xlabel('X (m)')
    ax.set_ylabel('Y (m)')
    ax.set_zlabel('Z (m)')
    if titles:
        ax.set_title('Body and Foot Trajectories')

    # Plot 2: Linear States
    fig2 = plt.figure(linear_state_fig)
    if linear_state_fig not in plt.get_fignums():
        fig2.set_size_inches(12, 6)
    fig2.clf()
    fig2.canvas.manager.set_window_title("linear_states")

    ax1 = plt.subplot(2, 1, 1)
    if "position" in state_traj and state_traj["position"] is not None:
        for i in range(3):
            ax1.plot(state_traj['time'], state_traj['position'][:, i],
                     color=body_color_vector[i], linestyle=line_style)

    ax1.set_ylabel('Position (m)')
    if titles:
        ax1.set_title('Linear States')
    ax1.legend(['X', 'Y', 'Z'], loc='right')
    ax1.autoscale(enable=True, axis='both', tight=True)
    ax1.grid(True)

    # Second subplot - Velocity
    ax2 = plt.subplot(2, 1, 2)
    if "velocity" in state_traj and state_traj["velocity"] is not None:
        for i in range(3):
            ax2.plot(state_traj['time'], state_traj['velocity'][:, i],
                     color=body_color_vector[i], linestyle=line_style)

    ax2.set_ylabel('Velocity (m/s)')
    ax2.set_xlabel('Time (s)')
    ax2.autoscale(enable=True, axis='both', tight=True)
    ax2.grid(True)

    # Align y-labels
    fig2.align_ylabels([ax1, ax2])
    plt.tight_layout()

    # Plot 3: Angular States
    if ("orientationRPY" in state_traj and
            state_traj["orientationRPY"] is not None):
        fig3 = plt.figure(angular_state_fig)
        if angular_state_fig not in plt.get_fignums():
            fig3.set_size_inches(12, 6)
        fig3.clf()
        fig3.canvas.manager.set_window_title("angular_states")

        ax1 = plt.subplot(2, 1, 1)
        for i in range(3):
            ax1.plot(state_traj['time'], state_traj['orientationRPY'][:, i],
                     color=body_color_vector[i], linestyle=line_style)

        ax1.set_ylabel('Ang. Pos. (rad)')
        if titles:
            ax1.set_title('Angular States')
        ax1.legend(['Roll', 'Pitch', 'Yaw'], loc='right')
        ax1.set_ylim([-pi, pi])
        ax1.autoscale(enable=True, axis='x', tight=True)
        ax1.grid(True)

        # Second subplot - Angular Velocity
        ax2 = plt.subplot(2, 1, 2)
        if ("angularVelocity" in state_traj and
                state_traj["angularVelocity"] is not None):
            for i in range(3):
                ax2.plot(state_traj['time'],
                         state_traj['angularVelocity'][:, i],
                         color=body_color_vector[i], linestyle=line_style)

        ax2.set_ylabel('Ang. Vel. (rad/s)')
        ax2.set_xlabel('Time (s)')
        ax2.autoscale(enable=True, axis='both', tight=True)
        ax2.grid(True)

        fig3.align_ylabels([ax1, ax2])
        plt.tight_layout()

    # Plot 4: Joint Positions
    if "jointPosition" in state_traj and state_traj["jointPosition"] is not None:
        # Specify indices for leg joints
        ab_index = [0, 3, 6, 9]
        hip_index = [1, 4, 7, 10]
        knee_index = [2, 5, 8, 11]
        abad_lim = [-1, 1]
        hip_lim = [-pi/2, pi]
        knee_lim = [0, pi]

        fig4 = plt.figure(joint_position_fig)
        if joint_position_fig not in plt.get_fignums():
            fig4.set_size_inches(12, 9)
        fig4.clf()
        fig4.canvas.manager.set_window_title("joint_positions")

        ax1 = plt.subplot(3, 1, 1)
        for i in range(num_feet):
            ax1.plot(state_traj['time'],
                     state_traj['jointPosition'][:, ab_index[i]],
                     color=joint_color_vector[i], linewidth=2,
                     linestyle=line_style)

        ax1.set_ylabel('Ab/Ad (rad)')
        if titles:
            ax1.set_title('Joint Angles')
        ax1.set_ylim(abad_lim)
        ax1.autoscale(enable=True, axis='x', tight=True)
        ax1.grid(True)

        ax2 = plt.subplot(3, 1, 2)
        for i in range(num_feet):
            ax2.plot(state_traj['time'],
                     state_traj['jointPosition'][:, hip_index[i]],
                     color=joint_color_vector[i], linewidth=2,
                     linestyle=line_style)

        ax2.set_ylabel('Hip (rad)')
        ax2.legend(['FL', 'BL', 'FR', 'BR'], loc='right')
        ax2.set_ylim(hip_lim)
        ax2.autoscale(enable=True, axis='x', tight=True)
        ax2.grid(True)

        ax3 = plt.subplot(3, 1, 3)
        for i in range(num_feet):
            ax3.plot(state_traj['time'],
                     state_traj['jointPosition'][:, knee_index[i]],
                     color=joint_color_vector[i], linewidth=2,
                     linestyle=line_style)

        ax3.set_ylabel('Knee (rad)')
        ax3.set_xlabel('Time (s)')
        ax3.set_ylim(knee_lim)
        ax3.autoscale(enable=True, axis='x', tight=True)
        ax3.grid(True)

        fig4.align_ylabels([ax1, ax2, ax3])
        plt.tight_layout()

    # Plot 5: Joint Velocities
    if "jointVelocity" in state_traj and state_traj["jointVelocity"] is not None:
        abad_vel_lim = [-37.7, 37.7]
        hip_vel_lim = [-37.7, 37.7]
        knee_vel_lim = [-25, 25]

        fig5 = plt.figure(joint_velocity_fig)
        if joint_velocity_fig not in plt.get_fignums():
            fig5.set_size_inches(12, 9)
        fig5.clf()
        fig5.canvas.manager.set_window_title("joint_velocities")

        ax1 = plt.subplot(3, 1, 1)
        for i in range(num_feet):
            ax1.plot(state_traj['time'],
                     state_traj['jointVelocity'][:, ab_index[i]],
                     color=joint_color_vector[i], linewidth=2,
                     linestyle=line_style)

        ax1.set_ylabel('Ab/Ad (rad/s)')
        if titles:
            ax1.set_title('Joint Velocities')
        ax1.set_ylim(abad_vel_lim)
        ax1.autoscale(enable=True, axis='x', tight=True)
        ax1.grid(True)

        ax2 = plt.subplot(3, 1, 2)
        for i in range(num_feet):
            ax2.plot(state_traj['time'],
                     state_traj['jointVelocity'][:, hip_index[i]],
                     color=joint_color_vector[i], linewidth=2,
                     linestyle=line_style)

        ax2.set_ylabel('Hip (rad/s)')
        ax2.legend(['FL', 'BL', 'FR', 'BR'], loc='right')
        ax2.set_ylim(hip_vel_lim)
        ax2.autoscale(enable=True, axis='x', tight=True)
        ax2.grid(True)

        ax3 = plt.subplot(3, 1, 3)
        for i in range(num_feet):
            ax3.plot(state_traj['time'],
                     state_traj['jointVelocity'][:, knee_index[i]],
                     color=joint_color_vector[i], linewidth=2,
                     linestyle=line_style)

        ax3.set_ylabel('Knee (rad/s)')
        ax3.set_xlabel('Time (s)')
        ax3.set_ylim(knee_vel_lim)
        ax3.autoscale(enable=True, axis='x', tight=True)
        ax3.grid(True)

        fig5.align_ylabels([ax1, ax2, ax3])
        plt.tight_layout()

    # Plot 6: Joint Effort
    if "jointEffort" in state_traj and state_traj["jointEffort"] is not None:
        abad_eff_lim = [-21, 21]
        hip_eff_lim = [-21, 21]
        knee_eff_lim = [-32, 32]

        fig6 = plt.figure(joint_effort_fig)
        if joint_effort_fig not in plt.get_fignums():
            fig6.set_size_inches(12, 9)
        fig6.clf()
        fig6.canvas.manager.set_window_title("joint_effort")

        ax1 = plt.subplot(3, 1, 1)
        for i in range(num_feet):
            ax1.plot(state_traj['time'],
                     state_traj['jointEffort'][:, ab_index[i]],
                     color=joint_color_vector[i], linewidth=2,
                     linestyle=line_style)

        ax1.set_ylabel('Ab/Ad (A)')
        if titles:
            ax1.set_title('Joint Effort')
        ax1.set_ylim(abad_eff_lim)
        ax1.autoscale(enable=True, axis='x', tight=True)
        ax1.grid(True)

        ax2 = plt.subplot(3, 1, 2)
        for i in range(num_feet):
            ax2.plot(state_traj['time'],
                     state_traj['jointEffort'][:, hip_index[i]],
                     color=joint_color_vector[i], linewidth=2,
                     linestyle=line_style)

        ax2.set_ylabel('Hip (A)')
        ax2.legend(['FL', 'BL', 'FR', 'BR'], loc='right')
        ax2.set_ylim(hip_eff_lim)
        ax2.autoscale(enable=True, axis='x', tight=True)
        ax2.grid(True)

        ax3 = plt.subplot(3, 1, 3)
        for i in range(num_feet):
            ax3.plot(state_traj['time'],
                     state_traj['jointEffort'][:, knee_index[i]],
                     color=joint_color_vector[i], linewidth=2,
                     linestyle=line_style)

        ax3.set_ylabel('Knee (A)')
        ax3.set_xlabel('Time (s)')
        ax3.set_ylim(knee_eff_lim)
        ax3.autoscale(enable=True, axis='x', tight=True)
        ax3.grid(True)

        fig6.align_ylabels([ax1, ax2, ax3])
        plt.tight_layout()

    # Plot 7: Foot Positions
    if "footPosition" in state_traj and state_traj["footPosition"]:
        fig7 = plt.figure(foot_position_fig)
        if foot_position_fig not in plt.get_fignums():
            fig7.set_size_inches(12, 9)
        fig7.clf()
        fig7.canvas.manager.set_window_title("foot_positions")

        ax1 = plt.subplot(3, 1, 1)
        for i in range(num_feet):
            ax1.plot(state_traj['time'],
                     state_traj['footPosition'][i][:, 0],
                     color=foot_color_vector[i], linewidth=2,
                     linestyle=line_style)

        ax1.set_ylabel('X (m)')
        if titles:
            ax1.set_title('Foot Positions')
        ax1.autoscale(enable=True, axis='both', tight=True)
        ax1.grid(True)

        ax2 = plt.subplot(3, 1, 2)
        for i in range(num_feet):
            ax2.plot(state_traj['time'],
                     state_traj['footPosition'][i][:, 1],
                     color=foot_color_vector[i], linewidth=2,
                     linestyle=line_style)

        ax2.set_ylabel('Y (m)')
        ax2.legend(['FL', 'BL', 'FR', 'BR'], loc='right')
        ax2.autoscale(enable=True, axis='both', tight=True)
        ax2.grid(True)

        ax3 = plt.subplot(3, 1, 3)
        for i in range(num_feet):
            ax3.plot(state_traj['time'],
                     state_traj['footPosition'][i][:, 2],
                     color=foot_color_vector[i], linewidth=2,
                     linestyle=line_style)

        ax3.set_ylabel('Z (m)')
        ax3.set_xlabel('Time (s)')
        ax3.autoscale(enable=True, axis='both', tight=True)
        ax3.grid(True)

        fig7.align_ylabels([ax1, ax2, ax3])
        plt.tight_layout()

    # Plot 8: Foot Velocities
    if "footVelocity" in state_traj and state_traj["footVelocity"]:
        fig8 = plt.figure(foot_velocity_fig)
        if foot_velocity_fig not in plt.get_fignums():
            fig8.set_size_inches(12, 9)
        fig8.clf()
        fig8.canvas.manager.set_window_title("foot_velocities")

        # Calculate velocity axis limits
        vel_axis = 0
        for i in range(num_feet):
            vel_axis = max(vel_axis,
                           np.max(np.abs(state_traj['footVelocity'][i])))

        ax1 = plt.subplot(3, 1, 1)
        for i in range(num_feet):
            ax1.plot(state_traj['time'],
                     state_traj['footVelocity'][i][:, 0],
                     color=foot_color_vector[i], linewidth=2,
                     linestyle=line_style)

        ax1.set_ylabel('X (m/s)')
        if titles:
            ax1.set_title('Foot Velocities')
        ax1.set_ylim([-vel_axis, vel_axis])
        ax1.autoscale(enable=True, axis='x', tight=True)
        ax1.grid(True)

        ax2 = plt.subplot(3, 1, 2)
        for i in range(num_feet):
            ax2.plot(state_traj['time'],
                     state_traj['footVelocity'][i][:, 1],
                     color=foot_color_vector[i], linewidth=2,
                     linestyle=line_style)

        ax2.set_ylabel('Y (m/s)')
        ax2.legend(['FL', 'BL', 'FR', 'BR'], loc='right')
        ax2.set_ylim([-vel_axis, vel_axis])
        ax2.autoscale(enable=True, axis='x', tight=True)
        ax2.grid(True)

        ax3 = plt.subplot(3, 1, 3)
        for i in range(num_feet):
            ax3.plot(state_traj['time'],
                     state_traj['footVelocity'][i][:, 2],
                     color=foot_color_vector[i], linewidth=2,
                     linestyle=line_style)

        ax3.set_ylabel('Z (m/s)')
        ax3.set_xlabel('Time (s)')
        ax3.set_ylim([-vel_axis, vel_axis])
        ax3.autoscale(enable=True, axis='x', tight=True)
        ax3.grid(True)

        fig8.align_ylabels([ax1, ax2, ax3])
        plt.tight_layout()

    # Return figure handles as numpy array
    fig_array = np.array([com_traj_fig, linear_state_fig, angular_state_fig,
                          joint_position_fig, joint_velocity_fig,
                          joint_effort_fig, foot_position_fig,
                          foot_velocity_fig])

    return fig_array
